zerver/lib/actions.py

```diff
@@ -5475,17 +5475,17 @@ def do_update_message_flags(
         if flag != "starred":
             raise JsonableError(_("Invalid message(s)"))
         # Validate that the user could have read the relevant message
-        message = access_message(user_profile, messages[0])[0]
-
-        # OK, this is a message that you legitimately have access
-        # to via narrowing to the stream it is on, even though you
-        # didn't actually receive it.  So we create a historical,
-        # read UserMessage message row for you to star.
-        UserMessage.objects.create(
-            user_profile=user_profile,
-            message=message,
-            flags=UserMessage.flags.historical | UserMessage.flags.read,
-        )
+        with transaction.atomic():
+            message = access_message(user_profile, messages[0], lock_message=True)[0]
+            # OK, this is a message that you legitimately have access
+            # to via narrowing to the stream it is on, even though you
+            # didn't actually receive it.  So we create a historical,
+            # read UserMessage message row for you to star.
+            UserMessage.objects.update_or_create(
+                user_profile=user_profile,
+                message=message,
+                flags=UserMessage.flags.historical | UserMessage.flags.read,
```

Comment:
Is the right way to call `update_or_create`?  I think you need to rearrange how `flags` are passed for it, but am not sure.

Reply:
The `update` case doesn't really do anything here (other than avoiding `IntegrityErrors` from duplicate entries), because if the reaction code had created a UserMessage in the meanwhile, it would be the same as what this codepath would have created.

Reply:
Update: I have made this code more explicit by doing a `.exists()` check, and also adding a comment to explain in what situations there may be a `UserMessage` already present.